### PR TITLE
Allow custom stages in aggregation pipeline

### PR DESCRIPTION
**Pull request by richarddd**
This will let users add more stages

### CODE REVIEW

File: core/src/main/java/dev/morphia/aggregation/Aggregation.java [hunk 1]
```diff
@@ -1,32 +1,13 @@
...
-import dev.morphia.aggregation.stages.AddFields;
-import dev.morphia.aggregation.stages.AutoBucket;
-import dev.morphia.aggregation.stages.Bucket;
-import dev.morphia.aggregation.stages.ChangeStream;
-import dev.morphia.aggregation.stages.CollectionStats;
-import dev.morphia.aggregation.stages.CurrentOp;
-import dev.morphia.aggregation.stages.Densify;
-import dev.morphia.aggregation.stages.Facet;
-import dev.morphia.aggregation.stages.Fill;
+import dev.morphia.aggregation.stages.*;
```
Comment (evanchooly) [inline core/src/main/java/dev/morphia/aggregation/Aggregation.java:7]: * imports break checkstyle so you'll want to clean that up.  The code formatter that runs as part of the maven build *should* clean that up but if not, you'll need to manually expand those.

File: core/src/main/java/dev/morphia/aggregation/Aggregation.java [hunk 2]
```diff
@@ -497,20 +478,23 @@ default Aggregation<T> set(AddFields fields) {
...
-     *
      * @return this
-     * @since 2.3
      * @aggregation.expression $changeStream
+     * @since 2.3
      */
-    Aggregation changeStream(ChangeStream stream);
+    Aggregation<T> changeStream(ChangeStream stream);
+
+    List<Stage> getStates();
```
Comment (evanchooly) [inline core/src/main/java/dev/morphia/aggregation/Aggregation.java:497]: This method should not be on the public interface.  I understand *why* it's there.  You want to test.  But you should downcast in your test.  There's no reason for user code to call this as it's an internal implementation detail.

File: core/src/main/java/dev/morphia/aggregation/AggregationImpl.java [hunk 3]
```diff
@@ -113,44 +82,37 @@ public AggregationImpl(Datastore datastore, Class<T> source, MongoCollection<T>
 
     @Override
     public Aggregation<T> autoBucket(AutoBucket bucket) {
-        addStage(bucket);
-        return this;
+        return addStage(bucket);
```
Comment (evanchooly) [inline core/src/main/java/dev/morphia/aggregation/AggregationImpl.java:85]: These changes are a little noisy in this PR.  But i get it.  I'd been meaning to do this but it just never bothered me enough to clean it up.

File: core/src/main/java/dev/morphia/aggregation/AggregationImpl.java [hunk 4]
```diff
@@ -270,110 +223,97 @@ public <O> void out(Out<O> out, AggregationOptions options) {
...
     @Override
-    public Aggregation changeStream(ChangeStream stream) {
-        addStage(stream);
-        return this;
+    public Aggregation<T> changeStream(ChangeStream stream) {
+        return addStage(stream);
+    }
+
+    @Override
+    public List<Stage> getStates() {
```
Comment (evanchooly) [inline core/src/main/java/dev/morphia/aggregation/AggregationImpl.java:315]: typo in the method name

**Review by evanchooly — CHANGES_REQUESTED** [review]
Thanks for the PR.  I'd been weighing a similar change.  Note that 2.3, when it comes out soon, should have all the stages implemented but I know sometimes a Morphia release can lag behind the release of new pipeline stages in the server so this will be a nice stopgap for the industrious until Morphia can catch up.

There's a little clean up to do here and there but once that's done we can get this merged.

**Comment by evanchooly** [comment]
@richarddd any updates?

**Comment by richarddd** [comment]
> @richarddd any updates?

Done!

**Comment by evanchooly** [comment]
@richarddd do you understand what needs to be fixed on the PR?

**Comment by richarddd** [comment]
> @richarddd do you understand what needs to be fixed on the PR?

Not quite? I did change what you've requested and updated it again

**Comment by evanchooly** [comment]
There are some style checks failing.  You can see in the message at the end how to fix it.  you can run those checks locally with `mvn -Dcode-audits` and you'll see the error, and fix, when it's done.

**Comment by evanchooly** [comment]
i applied the audit fixes for you but I need you to rebase your PR.  It looks like you have older commits in your log.

**Comment by evanchooly** [comment]
manually merged

**Pull request closed without merging**